`missing client_email` when creating supply client with workload identity federation token

### New Issue Checklist

- [x] Updated fastlane to the latest version
- [x] I read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [x] I read [docs.fastlane.tools](https://docs.fastlane.tools)
- [x] I searched for [existing GitHub issues](https://github.com/fastlane/fastlane/issues)

### Issue Description

I am trying to build and deploy an Android app using GitHub Action, with [credentials obtained through Workflow Identity Federation](https://github.com/google-github-actions/auth). When making requests to Google Play, `supply` throw an error: `missing client_email`.

I am opting for credential generated through Workflow Identity Federation instead of JSON credentials file because the JSON credentials never expires. Using federated identity helps rotate credentials and reduce number of sensitivity materials stored in GitHub Action.

The credentials JSON obtained through Workflow Identity Federation looks like this:

<pre>
{
  "type": "external_account",
  "audience": "//iam.googleapis.com/projects/ftl-test-app/locations/global/workloadIdentityPools/identity-pool/providers/token-provider",
  "subject_token_type": "urn:ietf:params:oauth:token-type:jwt",
  "token_url": "https://sts.googleapis.com/v1/token",
  "credential_source": {
    "url": "https://credential-source.example.com/token",
    "format": { "type": "json", "subject_token_field_name": "value" }
  },
  "service_account_impersonation_url": "https://iamcredentials.googleapis.com/v1/projects/-/serviceAccounts/fastlane@fastlane-tools.iam.gserviceaccount.com:generateAccessToken"
}
</pre>

##### Command executed

```
bundle exec fastlane android deploy
```

##### Complete output when running fastlane, including the stack trace and command used


<details>
  <pre>
exec fastlane android deploy
  
+---------------------------------------------------------------------------+
|                               Used plugins                                |
+----------------------------------------+---------+------------------------+
| Plugin                                 | Version | Action                 |
+----------------------------------------+---------+------------------------+
| fastlane-plugin-increment_version_code | 0.4.3   | increment_version_code |
| fastlane-plugin-increment_version_name | 0.0.10  | increment_version_name |
+----------------------------------------+---------+------------------------+

[08:16:20]: Sending anonymous analytics information
[08:16:20]: Learn more at https://docs.fastlane.tools/#metrics
[08:16:20]: No personal or sensitive data is sent.
[08:16:20]: You can disable this by adding `opt_out_usage` at the top of your Fastfile
[08:16:20]: ------------------------------
[08:16:20]: --- Step: default_platform ---
[08:16:20]: ------------------------------
[08:16:20]: Driving the lane 'android deploy' 🚀
[08:16:20]: ----------------------------------------------
[08:16:20]: --- Step: Switch to android increment lane ---
[08:16:20]: ----------------------------------------------
[08:16:20]: Cruising over to lane 'android increment' 🚖
[08:16:20]: ---------------------------------------------
[08:16:20]: --- Step: google_play_track_version_codes ---
[08:16:20]: ---------------------------------------------
+-----------------------------------+
|           Lane Context            |
+------------------+----------------+
| DEFAULT_PLATFORM | android        |
| PLATFORM_NAME    | android        |
| LANE_NAME        | android deploy |
+------------------+----------------+
[08:16:20]: Called from Fastfile at line 36
[08:16:20]: ```
[08:16:20]:     34:	  desc "Increment version code"
[08:16:20]:     35:	  lane :increment do
[08:16:20]:  => 36:	    version_code = google_play_track_version_codes[0] + 1
[08:16:20]:     37:	    increment_version_code(version_code: version_code)
[08:16:20]:     38:	  end
[08:16:20]: ```
[08:16:20]: missing client_email

+-------------------------------------------------------+
|                   fastlane summary                    |
+------+----------------------------------+-------------+
| Step | Action                           | Time (in s) |
+------+----------------------------------+-------------+
| 1    | default_platform                 | 0           |
| 2    | Switch to android increment lane | 0           |
| 💥   | google_play_track_version_codes  | 0           |
+------+----------------------------------+-------------+

[08:16:20]: fastlane finished with errors

Looking for related GitHub issues on fastlane/fastlane...

➡️  Validate Credential Account
    https://github.com/fastlane/fastlane/pull/22122 [open] 2 💬
    01 Aug 2024

➡️  Validate credential account
    https://github.com/fastlane/fastlane/pull/22120 [closed] 0 💬
    03 Jul 2024

➡️  (supply): Could not find option 'key' in the list of available options: json_key, json_key_data, root_url, timeout
    https://github.com/fastlane/fastlane/issues/18338 [closed] 1 💬
    05 May 2021

and 3 more at: https://github.com/fastlane/fastlane/search?q=missing%20client_email&type=Issues&utf8=✓

bundler: failed to load command: fastlane (/home/runner/work/repo/repo/android/vendor/bundle/ruby/3.2.0/bin/fastlane)
/home/runner/work/repo/repo/android/vendor/bundle/ruby/3.2.0/gems/googleauth-1.8.1/lib/googleauth/json_key_reader.rb:24:in `read_json_key': [!] missing client_email (RuntimeError)
	from /home/runner/work/repo/repo/android/vendor/bundle/ruby/3.2.0/gems/googleauth-1.8.1/lib/googleauth/service_account.rb:56:in `make_creds'
	from /home/runner/work/repo/repo/android/vendor/bundle/ruby/3.2.0/gems/fastlane-2.227.0/supply/lib/supply/client.rb:49:in `initialize'
	from /home/runner/work/repo/repo/android/vendor/bundle/ruby/3.2.0/gems/fastlane-2.227.0/supply/lib/supply/client.rb:19:in `new'
	from /home/runner/work/repo/repo/android/vendor/bundle/ruby/3.2.0/gems/fastlane-2.227.0/supply/lib/supply/client.rb:19:in `make_from_config'
	from /home/runner/work/repo/repo/android/vendor/bundle/ruby/3.2.0/gems/fastlane-2.227.0/supply/lib/supply/reader.rb:38:in `client'
	from /home/runner/work/repo/repo/android/vendor/bundle/ruby/3.2.0/gems/fastlane-2.227.0/supply/lib/supply/reader.rb:6:in `track_version_codes'
	from /home/runner/work/repo/repo/android/vendor/bundle/ruby/3.2.0/gems/fastlane-2.227.0/fastlane/lib/fastlane/actions/google_play_track_version_codes.rb:26:in `run'
	from /home/runner/work/repo/repo/android/vendor/bundle/ruby/3.2.0/gems/fastlane-2.227.0/fastlane/lib/fastlane/runner.rb:263:in `block (2 levels) in execute_action'
	from /home/runner/work/repo/repo/android/vendor/bundle/ruby/3.2.0/gems/fastlane-2.227.0/fastlane/lib/fastlane/actions/actions_helper.rb:69:in `execute_action'
	from /home/runner/work/repo/repo/android/vendor/bundle/ruby/3.2.0/gems/fastlane-2.227.0/fastlane/lib/fastlane/runner.rb:255:in `block in execute_action'
	from /home/runner/work/repo/repo/android/vendor/bundle/ruby/3.2.0/gems/fastlane-2.227.0/fastlane/lib/fastlane/runner.rb:229:in `chdir'
	from /home/runner/work/repo/repo/android/vendor/bundle/ruby/3.2.0/gems/fastlane-2.227.0/fastlane/lib/fastlane/runner.rb:229:in `execute_action'
	from /home/runner/work/repo/repo/android/vendor/bundle/ruby/3.2.0/gems/fastlane-2.227.0/fastlane/lib/fastlane/runner.rb:157:in `trigger_action_by_name'
	from /home/runner/work/repo/repo/android/vendor/bundle/ruby/3.2.0/gems/fastlane-2.227.0/fastlane/lib/fastlane/fast_file.rb:159:in `method_missing'
	from Fastfile:36:in `block (2 levels) in parsing_binding'
	from /home/runner/work/repo/repo/android/vendor/bundle/ruby/3.2.0/gems/fastlane-2.227.0/fastlane/lib/fastlane/lane.rb:41:in `call'
	from /home/runner/work/repo/repo/android/vendor/bundle/ruby/3.2.0/gems/fastlane-2.227.0/fastlane/lib/fastlane/runner.rb:204:in `try_switch_to_lane'
	from /home/runner/work/repo/repo/android/vendor/bundle/ruby/3.2.0/gems/fastlane-2.227.0/fastlane/lib/fastlane/runner.rb:146:in `trigger_action_by_name'
	from /home/runner/work/repo/repo/android/vendor/bundle/ruby/3.2.0/gems/fastlane-2.227.0/fastlane/lib/fastlane/fast_file.rb:159:in `method_missing'
	from Fastfile:47:in `block (2 levels) in parsing_binding'
	from /home/runner/work/repo/repo/android/vendor/bundle/ruby/3.2.0/gems/fastlane-2.227.0/fastlane/lib/fastlane/lane.rb:41:in `call'
	from /home/runner/work/repo/repo/android/vendor/bundle/ruby/3.2.0/gems/fastlane-2.227.0/fastlane/lib/fastlane/runner.rb:49:in `block in execute'
	from /home/runner/work/repo/repo/android/vendor/bundle/ruby/3.2.0/gems/fastlane-2.227.0/fastlane/lib/fastlane/runner.rb:45:in `chdir'
	from /home/runner/work/repo/repo/android/vendor/bundle/ruby/3.2.0/gems/fastlane-2.227.0/fastlane/lib/fastlane/runner.rb:45:in `execute'
	from /home/runner/work/repo/repo/android/vendor/bundle/ruby/3.2.0/gems/fastlane-2.227.0/fastlane/lib/fastlane/lane_manager.rb:46:in `cruise_lane'
	from /home/runner/work/repo/repo/android/vendor/bundle/ruby/3.2.0/gems/fastlane-2.227.0/fastlane/lib/fastlane/command_line_handler.rb:34:in `handle'
	from /home/runner/work/repo/repo/android/vendor/bundle/ruby/3.2.0/gems/fastlane-2.227.0/fastlane/lib/fastlane/commands_generator.rb:110:in `block (2 levels) in run'
	from /home/runner/work/repo/repo/android/vendor/bundle/ruby/3.2.0/gems/commander-4.6.0/lib/commander/command.rb:187:in `call'
	from /home/runner/work/repo/repo/android/vendor/bundle/ruby/3.2.0/gems/commander-4.6.0/lib/commander/command.rb:157:in `run'
	from /home/runner/work/repo/repo/android/vendor/bundle/ruby/3.2.0/gems/commander-4.6.0/lib/commander/runner.rb:444:in `run_active_command'
	from /home/runner/work/repo/repo/android/vendor/bundle/ruby/3.2.0/gems/fastlane-2.227.0/fastlane_core/lib/fastlane_core/ui/fastlane_runner.rb:124:in `run!'
	from /home/runner/work/repo/repo/android/vendor/bundle/ruby/3.2.0/gems/commander-4.6.0/lib/commander/delegates.rb:18:in `run!'
	from /home/runner/work/repo/repo/android/vendor/bundle/ruby/3.2.0/gems/fastlane-2.227.0/fastlane/lib/fastlane/commands_generator.rb:363:in `run'
	from /home/runner/work/repo/repo/android/vendor/bundle/ruby/3.2.0/gems/fastlane-2.227.0/fastlane/lib/fastlane/commands_generator.rb:43:in `start'
	from /home/runner/work/repo/repo/android/vendor/bundle/ruby/3.2.0/gems/fastlane-2.227.0/fastlane/lib/fastlane/cli_tools_distributor.rb:123:in `take_off'
	from /home/runner/work/repo/repo/android/vendor/bundle/ruby/3.2.0/gems/fastlane-2.227.0/bin/fastlane:23:in `<top (required)>'
	from /home/runner/work/repo/repo/android/vendor/bundle/ruby/3.2.0/bin/fastlane:25:in `load'
	from /home/runner/work/repo/repo/android/vendor/bundle/ruby/3.2.0/bin/fastlane:25:in `<top (required)>'
	from /opt/hostedtoolcache/Ruby/3.2.3/x64/lib/ruby/gems/3.2.0/gems/bundler-2.3.19/lib/bundler/cli/exec.rb:58:in `load'
	from /opt/hostedtoolcache/Ruby/3.2.3/x64/lib/ruby/gems/3.2.0/gems/bundler-2.3.19/lib/bundler/cli/exec.rb:58:in `kernel_load'
	from /opt/hostedtoolcache/Ruby/3.2.3/x64/lib/ruby/gems/3.2.0/gems/bundler-2.3.19/lib/bundler/cli/exec.rb:23:in `run'
	from /opt/hostedtoolcache/Ruby/3.2.3/x64/lib/ruby/gems/3.2.0/gems/bundler-2.3.19/lib/bundler/cli.rb:483:in `exec'
	from /opt/hostedtoolcache/Ruby/3.2.3/x64/lib/ruby/gems/3.2.0/gems/bundler-2.3.19/lib/bundler/vendor/thor/lib/thor/command.rb:27:in `run'
	from /opt/hostedtoolcache/Ruby/3.2.3/x64/lib/ruby/gems/3.2.0/gems/bundler-2.3.19/lib/bundler/vendor/thor/lib/thor/invocation.rb:127:in `invoke_command'
	from /opt/hostedtoolcache/Ruby/3.2.3/x64/lib/ruby/gems/3.2.0/gems/bundler-2.3.19/lib/bundler/vendor/thor/lib/thor.rb:392:in `dispatch'
	from /opt/hostedtoolcache/Ruby/3.2.3/x64/lib/ruby/gems/3.2.0/gems/bundler-2.3.19/lib/bundler/cli.rb:31:in `dispatch'
	from /opt/hostedtoolcache/Ruby/3.2.3/x64/lib/ruby/gems/3.2.0/gems/bundler-2.3.19/lib/bundler/vendor/thor/lib/thor/base.rb:485:in `start'
	from /opt/hostedtoolcache/Ruby/3.2.3/x64/lib/ruby/gems/3.2.0/gems/bundler-2.3.19/lib/bundler/cli.rb:25:in `start'
	from /opt/hostedtoolcache/Ruby/3.2.3/x64/lib/ruby/gems/3.2.0/gems/bundler-2.3.19/exe/bundle:48:in `block in <top (required)>'
	from /opt/hostedtoolcache/Ruby/3.2.3/x64/lib/ruby/gems/3.2.0/gems/bundler-2.3.19/lib/bundler/friendly_errors.rb:120:in `with_friendly_errors'
	from /opt/hostedtoolcache/Ruby/3.2.3/x64/lib/ruby/gems/3.2.0/gems/bundler-2.3.19/exe/bundle:36:in `<top (required)>'
	from /opt/hostedtoolcache/Ruby/3.2.3/x64/bin/bundle:25:in `load'
	from /opt/hostedtoolcache/Ruby/3.2.3/x64/bin/bundle:25:in `<main>'
Error: Process completed with exit code 1.
</pre>
</details>

### Environment

<details><summary>✅ fastlane environment ✅</summary>

### Stack

| Key                         | Value                                                                  |
| --------------------------- | ---------------------------------------------------------------------- |
| OS                          | debian_based trixie/sid                                                |
| Ruby                        | 3.2.3                                                                  |
| Bundler?                    | true                                                                   |
| Git                         | git version 2.48.1                                                     |
| Installation Source         | ~/work/earzz-fe/earzz-fe/android/vendor/bundle/ruby/3.2.0/bin/fastlane |
| Host                        | Ubuntu 24.04.2 LTS debian_based trixie/sid (6.8.0-1021-azure)          |
| Ruby Lib Dir                | /opt/hostedtoolcache/Ruby/3.2.3/x64/lib                                |
| OpenSSL Version             | OpenSSL 3.0.13 30 Jan 2024                                             |
| Is contained                | false                                                                  |
| Is homebrew                 | false                                                                  |
| Is installed via Fabric.app | false                                                                  |


### System Locale

| Variable | Value   |   |
| -------- | ------- | - |
| LANG     | C.UTF-8 | ✅ |
| LC_ALL   |         |   |
| LANGUAGE |         |   |


### fastlane files:

<details><summary>`./fastlane/Fastfile`</summary>

```ruby
default_platform(:android)

platform :android do
  before_all do
    ENV["SUPPLY_TRACK"] = "internal"
    ENV["APP_ENV"] = ENV["APP_ENV"] || "production"
  end

  desc "Increment version code"
  lane :increment do
    version_code = google_play_track_version_codes[0] + 1
    increment_version_code(version_code: version_code)
  end
end

```
</details>

### fastlane gems

| Gem      | Version | Update-Status |
| -------- | ------- | ------------- |
| fastlane | 2.227.0 | ✅ Up-To-Date  |


### Loaded fastlane plugins:

| Plugin                                 | Version | Update-Status |
| -------------------------------------- | ------- | ------------- |
| fastlane-plugin-increment_version_code | 0.4.3   | ✅ Up-To-Date  |
| fastlane-plugin-increment_version_name | 0.0.10  | ✅ Up-To-Date  |


<details><summary><b>Loaded gems</b></summary>

| Gem                                    | Version      |
| -------------------------------------- | ------------ |
| error_highlight                        | 0.5.1        |
| did_you_mean                           | 1.6.3        |
| syntax_suggest                         | 1.1.0        |
| bundler                                | 2.3.19       |
| pathname                               | 0.2.1        |
| rake                                   | 13.2.1       |
| base64                                 | 0.2.0        |
| nkf                                    | 0.2.0        |
| rexml                                  | 3.4.1        |
| CFPropertyList                         | 3.0.7        |
| public_suffix                          | 6.0.1        |
| addressable                            | 2.8.7        |
| artifactory                            | 3.0.17       |
| atomos                                 | 0.1.3        |
| aws-eventstream                        | 1.3.2        |
| aws-partitions                         | 1.1067.0     |
| aws-sigv4                              | 1.11.0       |
| jmespath                               | 1.6.2        |
| aws-sdk-core                           | 3.220.1      |
| aws-sdk-kms                            | 1.99.0       |
| aws-sdk-s3                             | 1.182.0      |
| babosa                                 | 1.0.4        |
| claide                                 | 1.1.0        |
| colored                                | 1.2          |
| colored2                               | 3.1.2        |
| highline                               | 2.0.3        |
| commander                              | 4.6.0        |
| declarative                            | 0.0.20       |
| digest-crc                             | 0.7.0        |
| domain_name                            | 0.6.20240107 |
| dotenv                                 | 2.8.1        |
| emoji_regex                            | 3.2.3        |
| excon                                  | 0.112.0      |
| faraday-em_http                        | 1.0.0        |
| faraday-em_synchrony                   | 1.0.0        |
| faraday-excon                          | 1.1.0        |
| faraday-httpclient                     | 1.0.1        |
| multipart-post                         | 2.4.1        |
| faraday-multipart                      | 1.1.0        |
| faraday-net_http                       | 1.0.2        |
| faraday-net_http_persistent            | 1.2.0        |
| faraday-patron                         | 1.0.0        |
| faraday-rack                           | 1.0.0        |
| faraday-retry                          | 1.0.3        |
| ruby2_keywords                         | 0.0.5        |
| faraday                                | 1.10.4       |
| http-cookie                            | 1.0.8        |
| faraday-cookie_jar                     | 0.0.7        |
| faraday_middleware                     | 1.2.1        |
| fastimage                              | 2.4.0        |
| sysrandom                              | 1.0.5        |
| fastlane-sirp                          | 1.0.0        |
| gh_inspector                           | 1.1.3        |
| jwt                                    | 2.10.1       |
| multi_json                             | 1.15.0       |
| os                                     | 1.1.4        |
| signet                                 | 0.19.0       |
| googleauth                             | 1.8.1        |
| mutex_m                                | 0.3.0        |
| httpclient                             | 2.9.0        |
| mini_mime                              | 1.1.5        |
| trailblazer-option                     | 0.1.2        |
| uber                                   | 0.1.0        |
| representable                          | 3.2.0        |
| retriable                              | 3.1.2        |
| google-apis-core                       | 0.11.3       |
| google-apis-androidpublisher_v3        | 0.54.0       |
| google-apis-playcustomapp_v1           | 0.13.0       |
| google-cloud-env                       | 1.6.0        |
| google-apis-iamcredentials_v1          | 0.17.0       |
| google-apis-storage_v1                 | 0.31.0       |
| google-cloud-errors                    | 1.5.0        |
| google-cloud-core                      | 1.8.0        |
| google-cloud-storage                   | 1.47.0       |
| json                                   | 2.10.2       |
| mini_magick                            | 4.13.2       |
| naturally                              | 2.2.1        |
| optparse                               | 0.6.0        |
| plist                                  | 3.7.2        |
| rubyzip                                | 2.4.1        |
| security                               | 0.1.5        |
| simctl                                 | 1.6.10       |
| terminal-notifier                      | 2.0.0        |
| unicode-display_width                  | 2.6.0        |
| terminal-table                         | 3.0.2        |
| tty-screen                             | 0.8.2        |
| tty-cursor                             | 0.7.1        |
| tty-spinner                            | 0.9.3        |
| word_wrap                              | 1.0.0        |
| nanaimo                                | 0.4.0        |
| xcodeproj                              | 1.27.0       |
| rouge                                  | 3.28.0       |
| xcpretty                               | 0.4.0        |
| xcpretty-travis-formatter              | 1.0.1        |
| fastlane-plugin-increment_version_code | 0.4.3        |
| fastlane-plugin-increment_version_name | 0.0.10       |
</details>


*generated on:* **2025-03-15**
</details>

--- a/supply/lib/supply/client.rb
+++ b/supply/lib/supply/client.rb
@@ -46,7 +46,7 @@ module Supply
     # Initializes the service and its auth_client using the specified information
     # @param service_account_json: The raw service account Json data
     def initialize(service_account_json: nil, params: nil)
-      auth_client = Google::Auth::ServiceAccountCredentials.make_creds(json_key_io: service_account_json, scope: self.class::SCOPE)
+      auth_client = Google::Auth::DefaultCredentials.make_creds(json_key_io: service_account_json, scope: self.class::SCOPE)
 
       UI.verbose("Fetching a new access token from Google...")
 

--- a/supply/lib/supply/client.rb
+++ b/supply/lib/supply/client.rb
@@ -47,7 +47,13 @@ module Supply
     # @param service_account_json: The raw service account Json data
     def initialize(service_account_json: nil, params: nil)
       # decode the json and check its type
-      google_credentials = JSON.parse(service_account_json.string)
+      begin
+        json_content = service_account_json.read
+        google_credentials = JSON.parse(json_content)
+        service_account_json.rewind
+      rescue JSON::ParserError
+        UI.user_error!("Invalid Google Credentials file provided - unable to parse json.")
+      end
 
       # use correct credential class based on type
       case google_credentials['type']
@@ -56,7 +62,7 @@ module Supply
       when "service_account"
         auth_client = Google::Auth::ServiceAccountCredentials.make_creds(json_key_io: service_account_json, scope: self.class::SCOPE)
       else
-        UI.user_error!("Invalid Google Credentials file provided.")
+        UI.user_error!("Invalid Google Credentials file provided - no credential type found.")
       end
 
       UI.verbose("Fetching a new access token from Google...")

--- a/supply/lib/supply/client.rb
+++ b/supply/lib/supply/client.rb
@@ -46,7 +46,18 @@ module Supply
     # Initializes the service and its auth_client using the specified information
     # @param service_account_json: The raw service account Json data
     def initialize(service_account_json: nil, params: nil)
-      auth_client = Google::Auth::DefaultCredentials.make_creds(json_key_io: service_account_json, scope: self.class::SCOPE)
+      # decode the json and check its type
+      google_credentials = JSON.parse(service_account_json.string)
+
+      # use correct credential class based on type
+      case google_credentials['type']
+      when "external_account"
+        auth_client = Google::Auth::ExternalAccount::Credentials.make_creds(json_key_io: service_account_json, scope: self.class::SCOPE)
+      when "service_account"
+        auth_client = Google::Auth::ServiceAccountCredentials.make_creds(json_key_io: service_account_json, scope: self.class::SCOPE)
+      else
+        UI.user_error!("Invalid Google Credentials file provided.")
+      end
 
       UI.verbose("Fetching a new access token from Google...")
 

--- a/supply/spec/client_spec.rb
+++ b/supply/spec/client_spec.rb
@@ -1,11 +1,25 @@
 describe Supply do
   describe Supply::Client do
     let(:service_account_file) { File.read(fixture_file("sample-service-account.json")) }
+    let(:external_account_file) { File.read(fixture_file("sample-external-account.json")) }
     before do
       stub_request(:post, "https://www.googleapis.com/oauth2/v4/token").
         to_return(status: 200, body: '{}', headers: { 'Content-Type' => 'application/json' })
       stub_request(:post, "https://androidpublisher.googleapis.com/androidpublisher/v3/applications/test-app/edits").
         to_return(status: 200, body: '{}', headers: { 'Content-Type' => 'application/json' })
+    end
+
+    describe "initialize client" do
+      it "with external account credentials" do
+        stub_request(:get, "https://credential-source.example.com/token").
+          to_return(status: 200, body: '{"value": "access-token"}', headers: { 'Content-Type' => 'application/json' })
+        stub_request(:post, "https://sts.googleapis.com/v1/token").
+          to_return(status: 200, body: '{}', headers: { 'Content-Type' => 'application/json' })
+        stub_request(:post, "https://iamcredentials.googleapis.com/v1/projects/-/serviceAccounts/fastlane@fastlane-tools.iam.gserviceaccount.com:generateAccessToken").
+          to_return(status: 200, body: '{}', headers: { 'Content-Type' => 'application/json' })
+
+        Supply::Client.new(service_account_json: StringIO.new(external_account_file), params: { timeout: 1 })
+      end
     end
 
     describe "displays error messages from the API" do

--- a/supply/spec/client_spec.rb
+++ b/supply/spec/client_spec.rb
@@ -42,6 +42,12 @@ describe Supply do
         }.to raise_error(FastlaneCore::Interface::FastlaneError, "Google Api Error: Invalid request - Ensure project settings are enabled.")
       end
 
+      it "displays error messages for invalid credentials" do
+        expect {
+          Supply::Client.new(service_account_json: StringIO.new("{}"), params: { timeout: 1 })
+        }.to raise_error(FastlaneCore::Interface::FastlaneError, "Invalid Google Credentials file provided.")
+      end
+
       it "with 5 retries" do
         stub_const("ENV", { 'SUPPLY_UPLOAD_MAX_RETRIES' => 5 })
 

--- a/supply/spec/client_spec.rb
+++ b/supply/spec/client_spec.rb
@@ -20,6 +20,18 @@ describe Supply do
 
         Supply::Client.new(service_account_json: StringIO.new(external_account_file), params: { timeout: 1 })
       end
+
+      it "with external account credentials from file" do
+        stub_request(:get, "https://credential-source.example.com/token").
+          to_return(status: 200, body: '{"value": "access-token"}', headers: { 'Content-Type' => 'application/json' })
+        stub_request(:post, "https://sts.googleapis.com/v1/token").
+          to_return(status: 200, body: '{}', headers: { 'Content-Type' => 'application/json' })
+        stub_request(:post, "https://iamcredentials.googleapis.com/v1/projects/-/serviceAccounts/fastlane@fastlane-tools.iam.gserviceaccount.com:generateAccessToken").
+          to_return(status: 200, body: '{}', headers: { 'Content-Type' => 'application/json' })
+
+        file_path = fixture_file("sample-external-account.json")
+        Supply::Client.make_from_config(params: { json_key: file_path, timeout: 1 })
+      end
     end
 
     describe "displays error messages from the API" do
@@ -42,10 +54,16 @@ describe Supply do
         }.to raise_error(FastlaneCore::Interface::FastlaneError, "Google Api Error: Invalid request - Ensure project settings are enabled.")
       end
 
-      it "displays error messages for invalid credentials" do
+      it "displays error messages for invalid json" do
+        expect {
+          Supply::Client.new(service_account_json: StringIO.new("{/asdf"), params: { timeout: 1 })
+        }.to raise_error(FastlaneCore::Interface::FastlaneError, "Invalid Google Credentials file provided - unable to parse json.")
+      end
+
+      it "displays error messages for missing credential type" do
         expect {
           Supply::Client.new(service_account_json: StringIO.new("{}"), params: { timeout: 1 })
-        }.to raise_error(FastlaneCore::Interface::FastlaneError, "Invalid Google Credentials file provided.")
+        }.to raise_error(FastlaneCore::Interface::FastlaneError, "Invalid Google Credentials file provided - no credential type found.")
       end
 
       it "with 5 retries" do

--- a/supply/spec/fixtures/sample-external-account.json
+++ b/supply/spec/fixtures/sample-external-account.json
@@ -1,0 +1,11 @@
+{
+  "type": "external_account",
+  "audience": "//iam.googleapis.com/projects/ftl-test-app/locations/global/workloadIdentityPools/identity-pool/providers/token-provider",
+  "subject_token_type": "urn:ietf:params:oauth:token-type:jwt",
+  "token_url": "https://sts.googleapis.com/v1/token",
+  "credential_source": {
+    "url": "https://credential-source.example.com/token",
+    "format": { "type": "json", "subject_token_field_name": "value" }
+  },
+  "service_account_impersonation_url": "https://iamcredentials.googleapis.com/v1/projects/-/serviceAccounts/fastlane@fastlane-tools.iam.gserviceaccount.com:generateAccessToken"
+}

